### ISSUE TITLE
All ubuntu and debian upgrades test from 4.4 to 4.5 failed, could not Could not setup Async I/O.

### ISSUE THREAD

**Issue by shoshan** [scylladb/scylla#8713]
All ubuntu and debian upgrades test  from 4.4 to 4.5 failed with the following error:

```

Command: 'sudo /usr/lib/scylla/scylla_setup --nic ens5 --disks /dev/nvme0n1,/dev/nvme0n2,/dev/nvme0n3 --no-io-setup  --swap-directory / '
Exit code: 1
Stdout:
Unpacking xfsprogs (4.20.0-1) ...
Setting up libreadline5:amd64 (5.2+dfsg-3+b13) ...
Setting up xfsprogs (4.20.0-1) ...
update-initramfs: deferring update (trigger activated)
Processing triggers for man-db (2.8.5-2) ...
Processing triggers for libc-bin (2.28-10) ...
Processing triggers for initramfs-tools (0.133+deb10u1) ...
update-initramfs: Generating /boot/initrd.img-4.19.0-16-cloud-amd64
Please upgrade to a newer kernel version.
see https://docs.scylladb.com/troubleshooting/error_messages/kb_fs_not_qualified_aio/ for details
Stderr:
are supported and installed on your system.
perl: warning: Falling back to a fallback locale ("C.UTF-8").
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
ERROR 2021-05-26 09:48:19,095 [shard 7] seastar - Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
```

test:  scylla-4.5/rolling-upgrade/rolling-upgrade-debian10-test#30: 2021-05-26
ScyllaDB version: 4.4.2-0.20210522.2b29568bf (-)
Target ScyllaDB repo: http://downloads.scylladb.com/unstable/scylla/branch-4.5/deb/unified/2021-05-25T11:58:36Z/scylladb-4.5/scylla.list

logs:
https://cloudius-jenkins-test.s3.amazonaws.com/51f557ee-f7ae-44fd-a21a-f358c946b79e/20210526_094858/db-cluster-51f557ee.zip
https://cloudius-jenkins-test.s3.amazonaws.com/51f557ee-f7ae-44fd-a21a-f358c946b79e/20210526_094858/sct-runner-51f557ee.zip

**Comment by slivne** [comment]
We are going to change this (back) to be static

**Comment by penberg** [comment]
@shoshan Fixes for this are now in 4.5. Let's re-test and close this issue.

**Comment by shoshan** [comment]
@penberg  scylla-4.5/rolling-upgrade/rolling-upgrade-debian10-test#33: 2021-05-30 22:26:32
has same error.
```
update-initramfs: Generating /boot/initrd.img-4.19.0-16-cloud-amd64
Please upgrade to a newer kernel version.
see https://docs.scylladb.com/troubleshooting/error_messages/kb_fs_not_qualified_aio/ for details
Stderr:
perl: warning: Falling back to a fallback locale ("C.UTF-8").
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
ERROR 2021-05-30 22:34:24,319 [shard 2] seastar - Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
ERROR 2021-05-30 22:34:24,319 [shard 7] seastar - Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
```

Logs:
https://jenkins.scylladb.com/job/scylla-4.5/job/rolling-upgrade/job/rolling-upgrade-debian10-test/33/
https://cloudius-jenkins-test.s3.amazonaws.com/4f0e5e5f-a881-4316-99f0-deb388600a0b/20210530_223510/sct-runner-4f0e5e5f.zip
https://cloudius-jenkins-test.s3.amazonaws.com/4f0e5e5f-a881-4316-99f0-deb388600a0b/20210530_223510/db-cluster-4f0e5e5f.zip

**Comment by penberg** [comment]
Hmm, perhaps 7445bfec86af84ca4814335ece66904041c79cf1 is not working. /cc @yaronkaikov @syuu1228

**Comment by syuu1228** [comment]
I'll check

**Comment by shoshan** [comment]
rolling-upgrade-ubuntu18.04-test#34 2021-05-31 22:48:17
failed with
```
invoke.exceptions.UnexpectedExit: Encountered a bad command exit code!
Command: 'sudo /usr/lib/scylla/scylla_setup --nic ens5 --disks /dev/nvme0n1,/dev/nvme0n2,/dev/nvme0n3 --no-io-setup  --swap-directory / '
Exit code: 1
Stdout:
Please upgrade to a newer kernel version.
see https://docs.scylladb.com/troubleshooting/error_messages/kb_fs_not_qualified_aio/ for details
Stderr:
ERROR 2021-05-31 22:54:07,730 [shard 6] seastar - Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
```

probably same root cause

**Comment by syuu1228** [comment]
scylla-kernel-conf package on this repo DOES NOT contain 99-scylla-aio.conf:
```
vagrant@vagrant:~$ dpkg -L scylla-kernel-conf
/.
/usr
/usr/lib
/usr/lib/sysctl.d
/usr/lib/sysctl.d/99-scylla-inotify.conf
/usr/lib/sysctl.d/99-scylla-sched.conf
/usr/lib/sysctl.d/99-scylla-vm.conf
/usr/share
/usr/share/doc
/usr/share/doc/scylla-kernel-conf
/usr/share/doc/scylla-kernel-conf/changelog.Debian.gz
/usr/share/doc/scylla-kernel-conf/copyright
```
That's why we are getting error on the test.

So I thought it may debian packaging issue, but when I build debian package from latest branch-4.5, it DOES contain ​99-scylla-aio.conf:
```
[
[syuu@monster debian]$ dpkg -c scylla-kernel-conf_4.5~rc2-0.20210601.d92a26636-1_amd64.deb 
drwxr-xr-x root/root         0 2015-08-24 18:22 ./
drwxr-xr-x root/root         0 2015-08-24 18:22 ./usr/
drwxr-xr-x root/root         0 2015-08-24 18:22 ./usr/lib/
drwxr-xr-x root/root         0 2015-08-24 18:22 ./usr/lib/sysctl.d/
-rw-r--r-- root/root        47 2015-08-24 18:22 ./usr/lib/sysctl.d/99-scylla-aio.conf
-rw-r--r-- root/root       191 2015-08-24 18:22 ./usr/lib/sysctl.d/99-scylla-inotify.conf
-rw-r--r-- root/root       456 2015-08-24 18:22 ./usr/lib/sysctl.d/99-scylla-sched.conf
-rw-r--r-- root/root       340 2015-08-24 18:22 ./usr/lib/sysctl.d/99-scylla-vm.conf
drwxr-xr-x root/root         0 2015-08-24 18:22 ./usr/share/
drwxr-xr-x root/root         0 2015-08-24 18:22 ./usr/share/doc/
drwxr-xr-x root/root         0 2015-08-24 18:22 ./usr/share/doc/scylla-kernel-conf/
-rw-r--r-- root/root       164 2015-08-24 18:22 ./usr/share/doc/scylla-kernel-conf/changelog.Debian.gz
-rw-r--r-- root/root     38086 2015-08-24 18:22 ./usr/share/doc/scylla-kernel-conf/copyright
```

I'm still not sure why it's not configured properly on 2021-05-25T11:58:36Z but isn't it fixed on latest build?

**Comment by penberg** [comment]
Could this be a test issue? I see `sct_runner.py` is fiddling with `fs.aio-max-nr` tunable:

https://github.com/scylladb/scylla-cluster-tests/blob/master/sdcm/sct_runner.py#L76

Perhaps that's overriding the configuration we're setting?

/cc @syuu1228 @amoskong @yaronkaikov @shoshan

**Comment by penberg** [comment]
@syuu1228 When you say

> scylla-kernel-conf package on this repo DOES NOT contain 99-scylla-aio.conf:

_which_ repository are you talking about?

**Comment by yaronkaikov** [comment]
> Could this be a test issue? I see `sct_runner.py` is fiddling with `fs.aio-max-nr` tunable:
> 
> https://github.com/scylladb/scylla-cluster-tests/blob/master/sdcm/sct_runner.py#L76
> 
> Perhaps that's overriding the configuration we're setting?
> 
> /cc @syuu1228 @amoskong @yaronkaikov @shoshan

Well it looks like it

**Comment by roydahan** [comment]
> Could this be a test issue? I see `sct_runner.py` is fiddling with `fs.aio-max-nr` tunable:
> 
> https://github.com/scylladb/scylla-cluster-tests/blob/master/sdcm/sct_runner.py#L76
> 
> Perhaps that's overriding the configuration we're setting?
> 
> /cc @syuu1228 @amoskong @yaronkaikov @shoshan

I really doubt this is related, this code is in SCT_runner.
Which is not even used in GCE runs.

**Comment by roydahan** [comment]
It happens also in all deb/ubuntu master upgrade tests.

**Comment by syuu1228** [comment]
> It happens also in all deb/ubuntu master upgrade tests.

Hmm, I'll try to look into .deb builds on jenkins

**Comment by syuu1228** [comment]
> scylla-kernel-conf package on this repo DOES NOT contain 99-scylla-aio.conf:
> 
> ```
> vagrant@vagrant:~$ dpkg -L scylla-kernel-conf
> /.
> /usr
> /usr/lib
> /usr/lib/sysctl.d
> /usr/lib/sysctl.d/99-scylla-inotify.conf
> /usr/lib/sysctl.d/99-scylla-sched.conf
> /usr/lib/sysctl.d/99-scylla-vm.conf
> /usr/share
> /usr/share/doc
> /usr/share/doc/scylla-kernel-conf
> /usr/share/doc/scylla-kernel-conf/changelog.Debian.gz
> /usr/share/doc/scylla-kernel-conf/copyright
> ```
> 
> That's why we are getting error on the test.
> 
> So I thought it may debian packaging issue, but when I build debian package from latest branch-4.5, it DOES contain ​99-scylla-aio.conf:
> 
> ```
> [
> [syuu@monster debian]$ dpkg -c scylla-kernel-conf_4.5~rc2-0.20210601.d92a26636-1_amd64.deb 
> drwxr-xr-x root/root         0 2015-08-24 18:22 ./
> drwxr-xr-x root/root         0 2015-08-24 18:22 ./usr/
> drwxr-xr-x root/root         0 2015-08-24 18:22 ./usr/lib/
> drwxr-xr-x root/root         0 2015-08-24 18:22 ./usr/lib/sysctl.d/
> -rw-r--r-- root/root        47 2015-08-24 18:22 ./usr/lib/sysctl.d/99-scylla-aio.conf
> -rw-r--r-- root/root       191 2015-08-24 18:22 ./usr/lib/sysctl.d/99-scylla-inotify.conf
> -rw-r--r-- root/root       456 2015-08-24 18:22 ./usr/lib/sysctl.d/99-scylla-sched.conf
> -rw-r--r-- root/root       340 2015-08-24 18:22 ./usr/lib/sysctl.d/99-scylla-vm.conf
> drwxr-xr-x root/root         0 2015-08-24 18:22 ./usr/share/
> drwxr-xr-x root/root         0 2015-08-24 18:22 ./usr/share/doc/
> drwxr-xr-x root/root         0 2015-08-24 18:22 ./usr/share/doc/scylla-kernel-conf/
> -rw-r--r-- root/root       164 2015-08-24 18:22 ./usr/share/doc/scylla-kernel-conf/changelog.Debian.gz
> -rw-r--r-- root/root     38086 2015-08-24 18:22 ./usr/share/doc/scylla-kernel-conf/copyright
> ```
> 
> I'm still not sure why it's not configured properly on 2021-05-25T11:58:36Z but isn't it fixed on latest build?

Ah, my previous comment was wrong, http://downloads.scylladb.com/unstable/scylla/branch-4.5/deb/unified/2021-05-25T11:58:36Z/scylladb-4.5/scylla.list is point to 44f85d2, so it's before merging static aio.conf (7445bfe), so missing static aio.conf is correct.
On this version AIO parameter is dynamically configured by python scripts.
However, dynamic configuration script had a bug, it forgot to set AIO parameter before running scylla_io_setup (#8587).
We merged fix at , but 44f85d2 is older than that.
So 
```seastar - Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application```
is likely #8587, need to wait fix for later builds.

But it's strange if we keep failing on latest builds, because we merged fix at `May 27, 2021`.

**Comment by syuu1228** [comment]
On https://jenkins.scylladb.com/job/scylla-4.5/job/rolling-upgrade/job/rolling-upgrade-debian10-test/35/, it keep failing but list file (http://downloads.scylladb.com/unstable/scylla/branch-4.5/deb/unified/2021-06-02T14:37:15Z/scylladb-4.5/scylla.list) points to d92a26636, this version should contains all AIO fixes.

**Comment by syuu1228** [comment]
Note:

Seems like this error message was not scylla_io_setup, it is scylla_kernel_check.
But both are calling iotune, so It is same issue as #8587 anyway.

```
Please upgrade to a newer kernel version.
see https://docs.scylladb.com/troubleshooting/error_messages/kb_fs_not_qualified_aio/ for details
Stderr:
are supported and installed on your system.
ERROR 2021-05-26 09:48:19,095 [shard 7] seastar - Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
```

**Comment by syuu1228** [comment]
Tested Scylla 4.5 installation exactly same repository as https://jenkins.scylladb.com/job/scylla-4.5/job/rolling-upgrade/job/rolling-upgrade-debian10-test/35/ on newly created Ubuntu 20.04 VM, I do not see any problem:

```
vagrant@vagrant:/etc/apt/sources.list.d$ sudo wget http://downloads.scylladb.com/unstable/scylla/branch-4.5/deb/unified/2021-06-02T14:37:15Z/scylladb-4.5/scylla.list

(cut unrelated logs)

vagrant@vagrant:/etc/apt/sources.list.d$ sudo apt-get update --allow-insecure-repositories

(cut unrelated logs)

vagrant@vagrant:/etc/apt/sources.list.d$ sudo apt-get install scylla

(cut unrelated logs)

vagrant@vagrant:/etc/apt/sources.list.d$ sudo dpkg -l|grep scylla-server
ii  scylla-server                        4.5~rc2-0.20210602.d92a26636-1    amd64        Scylla database server binaries
vagrant@vagrant:/etc/apt/sources.list.d$ sysctl fs.aio-max-nr
fs.aio-max-nr = 5578536
vagrant@vagrant:/etc/apt/sources.list.d$ cat /usr/lib/sysctl.d/99-scylla-aio.conf 
# Raise max AIO events
fs.aio-max-nr = 5578536
vagrant@vagrant:/etc/apt/sources.list.d$ sudo scylla_setup --nic eth0 --disks /dev/sdb,/dev/sdc --no-io-setup --swap-directory /
WARN  2021-06-07 08:35:07,241 [shard 0] iotune - Available space on filesystem at /var/tmp/mnt: 121 MB: is less than recommended: 10 GB
INFO  2021-06-07 08:35:07,241 [shard 0] iotune - /var/tmp/mnt passed sanity checks
This is a supported kernel version.
Created symlink /etc/systemd/system/multi-user.target.wants/scylla-server.service → /lib/systemd/system/scylla-server.service.
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  libnss-systemd libpam-systemd libsystemd0 systemd systemd-sysv
Suggested packages:
  systemd-container
The following packages will be REMOVED:
  systemd-timesyncd
The following NEW packages will be installed:
  chrony
The following packages will be upgraded:
  libnss-systemd libpam-systemd libsystemd0 systemd systemd-sysv
5 upgraded, 1 newly installed, 1 to remove and 197 not upgraded.
Need to get 4,587 kB of archives.
After this operation, 534 kB of additional disk space will be used.

(cut unrelated logs)

ScyllaDB setup finished.
```

**Comment by syuu1228** [comment]
Okay, now I probably understood what's happening.
I think we have fs.aio-max-nr bug on **4.4**, not 4.5 (we had a bug on the version too, but fixed).
On 4.4.2, we had merged 8437f71 but no fixes backported after that, so it still has #8587, and fs.aio-max-nr = 65536 (distribution default).
This is not enough when cpus are many (probably it will cause error when ncpus > 6).
```
vagrant@vagrant:/etc/apt/sources.list.d$ dpkg -l|grep scylla-server
ii  scylla-server                        4.4.2-0.20210522.2b29568bf-1      amd64        Scylla database server binaries
vagrant@vagrant:/etc/apt/sources.list.d$ sysctl fs.aio-max-nr
fs.aio-max-nr = 65536
vagrant@vagrant:~$ lscpu|grep -e "^CPU(s):"
CPU(s):                          20
vagrant@vagrant:/etc/apt/sources.list.d$ sudo scylla_kernel_check
ERROR 2021-06-07 09:14:18,268 [shard 10] seastar - Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application
Please upgrade to a newer kernel version.
 see https://docs.scylladb.com/troubleshooting/error_messages/kb_fs_not_qualified_aio/ for details
```

I think we need to backport dd453ff and 588a065 to branch-4.4, just like branch-4.5.

**Comment by penberg** [comment]
@avikivity @slivne Do we want to backport the `aio-max-nr` changes to 4.4.x, or just mention this issue in 4.5.0 release notes and tell users to manually bump up the limit before upgrading?

**Comment by slivne** [comment]
@penberg  - decision we will do the backports

**Comment by bhalevy** [comment]
Given that #8845 is considered a duplicate of this bug, and it's already closed, anything prevents closing this issue?

**Comment by syuu1228** [comment]
@bhalevy Patches are backported to branch-4.4 at Jun 7, 2021, but Jenkins build won't fix until publishing new release of 4.4.x, because the test is installing 4.4 packages from 4.4 repo. Otherwise we need workaround on the test.
Also I think it's better to warn users to update 4.4.x packages to newest one, before upgrading 4.5 (probably in https://github.com/scylladb/scylla-docs/tree/master/upgrade/upgrade-opensource/upgrade-guide-from-4.4-to-4.5)

**Comment by slivne** [comment]
According to QA 4.4.3,4.3.5 4.5.rc3 are all fixed closing the issue
